app/model/repository/AssignmentSolutions.php

```diff
@@ -102,28 +182,39 @@ public function findBestSolutionsForAssignments(array $assignments, User $user):
       }
 
       $best = Arrays::get($result, $assignment->getId(), null);
+      $result[$assignment->getId()] = self::compareBestSolution($best, $solution);
+    }
 
-      if ($best === null) {
-        $result[$assignment->getId()] = $solution;
-        continue;
-      }
-
+    return $result;
+  }
 
-      if ($best->isAccepted()) {
+  /**
+   * Filter the given *complete* array of solutions so that only best solutions remain.
+   * Term *complete* means that if solution S of assignment A and user U is in the input set,
+   * all solutions from A by user U are also in the set.
+   * @param AssignmentSolution[] $solutions
+   * @return AssignmentSolution[]
+   */
+  public function filterBestSolutions(array $solutions) {
+    $result = [];
+    $validSolutions = self::filterValidSolutions($solutions);
+    foreach ($validSolutions as $solution) {
+      $assignment = $solution->getAssignment();
+      if ($assignment === null) {
         continue;
       }
+      $assignmentId = $assignment->getId();
 
-      if ($solution->isAccepted()) {
-        $result[$assignment->getId()] = $solution;
+      $author = $solution->getSolution()->getAuthor();
+      if ($author === null) {
         continue;
       }
+      $key = $assignment->getId() . ':' . $author->getId();
```

Comment:
You almost got me there :-) ... I almost wanted to point out that you do not have nested arrays...

Reply:
OK, I'll add a comment